include/nlohmann/detail/input/binary_reader.hpp

```diff
@@ -2179,7 +2179,7 @@ class binary_reader
                     for (auto i : dim)
                     {
                         result *= i;
-                        if (result == 0) // because dim elements shall not have zeros, result = 0 means overflow happened
+                        if (result == 0 || result == string_t::npos) // because dim elements shall not have zeros, result = 0 means overflow happened; it also can't be string_t::npos
```

Comment:
Don't you mean `static_cast<std::size_t>(-1)`? `string_t::npos` will probably be the same but with custom string types that's not guaranteed.

Reply:
Of course, we used that elsewhere and no one noticed that that's not 100% correct. It also predates your work.

Maybe leave it for now. We can decide separately whether to replace it with `static_cast<std::size_t>(-1)` or add a constant to the class.

Reply:
@falbrechtskirchinger, we have been assigning and comparing with `string_t::npos` in other parts of the code

https://github.com/nlohmann/json/blob/f6acdbec2c99670cf932d83cd797f08662e0b564/include/nlohmann/detail/input/binary_reader.hpp#L2228
https://github.com/nlohmann/json/blob/f6acdbec2c99670cf932d83cd797f08662e0b564/include/nlohmann/detail/input/binary_reader.hpp#L1949
https://github.com/nlohmann/json/blob/f6acdbec2c99670cf932d83cd797f08662e0b564/include/nlohmann/detail/input/binary_reader.hpp#L2488

also, seems `string_t::npos` is defined as the above form
https://cplusplus.com/reference/string/string/npos/

I am fine with either forms - if we switch, we need to make sure other places are also updated.

Reply:
I think having our own constant and setting it to `static_cast<std::size_t>(-1)` makes it clearer.

Reply:
`basic_json::string_t` isn't guaranteed to be `std::string`. In practice, it's unlikely to be a problem, it's just technically wrong.

I may fix that as part of my BJData performance PR. I haven't gotten around to finishing and submitting it.

Reply:
> I think having our own constant and setting it to `static_cast<std::size_t>(-1)` makes it clearer.

That's what I was going to do.

Reply:
> basic_json::string_t isn't guaranteed to be std::string. In practice, it's unlikely to be a problem, it's just technically wrong.

I see.

speaking of performance, I see for-loops are used to read UBJSON/BJData array elements of uniform types - from my experience in C (did not test this on C++), reading the entire buffer using something like `memcpy` would lead to good speed improvement (in the case of a BJData JS parser, using `new Uint8Array(Buffer.from() ...)` to [replace a for-loop](https://github.com/NeuroJSON/js-bjdata/commit/110a8de7348b90b88862fc2516b49ef191b92b8a#diff-e0039f8896faac0a839a0df36d7016c8bd78c665380d2e8d44a0495c9300eb1eR569) also result in dramatic speed up). 

https://github.com/nlohmann/json/blob/f6acdbec2c99670cf932d83cd797f08662e0b564/include/nlohmann/detail/input/binary_reader.hpp#L2538-L2550

Reply:
`get_number()` would be the place to apply such an optimization.

I'm talking about your use of `std::vector` and `std::map`, which incurs an unnecessary memory allocation. That will have some small but easily avoidable impact on performance, including heap fragmentation, which can be a problem with `glibc` on Linux.

Unfortunately, with C+11 the syntax for initializing a `std::array` is quite clunky compared to `std::vector`/`std::map` and I haven't submitted the PR yet for that reason.